source/beamline/SGM/energy/SGMEnergyTrajectory.cpp

```diff
@@ -1,199 +1,218 @@
 #include "SGMEnergyTrajectory.h"
+#include <math.h>
 
 SGMEnergyTrajectory::SGMEnergyTrajectory(double startEnergy,
                                          double endEnergy,
                                          double time,
-                                         SGMGratingSupport::GratingTranslation gratingTranslation)
+                                         SGMGratingSupport::GratingTranslation gratingTranslation,
+                                         double gratingAngleAcceleration,
+                                         double gratingAngleStepsPerEncoderCount,
+                                         double undulatorAcceleration,
+                                         double undulatorCurrentGap,
+                                         double undulatorCurrentStep)
 {
 
-    startEnergyPosition_ = new SGMEnergyPosition(startEnergy, gratingTranslation);
-    startEnergy_ = startEnergy;
-    endEnergyPosition_ = new SGMEnergyPosition(endEnergy, gratingTranslation);
-    endEnergy_ = endEnergy;
+	startEnergyPosition_ = new SGMEnergyPosition(startEnergy, gratingTranslation);
+	startEnergy_ = startEnergy;
+	endEnergyPosition_ = new SGMEnergyPosition(endEnergy, gratingTranslation);
+	endEnergy_ = endEnergy;
 
-    // ensure both undulator harmonics are the same.
-    if(startEnergyPosition_->undulatorHarmonic() != endEnergyPosition_->undulatorHarmonic()) {
+	errorValidator_ = new AMValidator();
+	errorValidator_->addChildValidator(startEnergyPosition_->errorValidator());
+	errorValidator_->addChildValidator(endEnergyPosition_->errorValidator());
 
-        // If there is a conflict, just use the first harmonic.
-        startEnergyPosition_->setAutoDetectUndulatorHarmonic(false);
-        endEnergyPosition_->setAutoDetectUndulatorHarmonic(false);
+	warningValidator_ = new AMValidator();
+	warningValidator_->addChildValidator(startEnergyPosition_->warningValidator());
+	warningValidator_->addChildValidator(endEnergyPosition_->warningValidator());
 
-        startEnergyPosition_->setUndulatorHarmonic(SGMUndulatorSupport::FirstHarmonic);
-        endEnergyPosition_->setUndulatorHarmonic(SGMUndulatorSupport::FirstHarmonic);
-    }
+	// ensure both undulator harmonics are the same.
+	if(startEnergyPosition_->undulatorHarmonic() != endEnergyPosition_->undulatorHarmonic()) {
 
-    time_ = time;
+		// If there is a conflict, just use the first harmonic.
+		startEnergyPosition_->setAutoDetectUndulatorHarmonic(false);
+		endEnergyPosition_->setAutoDetectUndulatorHarmonic(false);
+
+		startEnergyPosition_->setUndulatorHarmonic(SGMUndulatorSupport::FirstHarmonic);
+		endEnergyPosition_->setUndulatorHarmonic(SGMUndulatorSupport::FirstHarmonic);
+	}
+
+	time_ = time;
+
+	gratingAngleVelocityProfile_ = AMTrapezoidVelocityProfile(startEnergyPosition_->gratingAngle() * gratingAngleStepsPerEncoderCount,
```

Comment:
The conversion here is due to the fact that the acceleration and velocity are in steps but all of the positions are in real units, right?